Zend/zend_fibers.c

```diff
@@ -0,0 +1,778 @@
+
+	zend_object_std_dtor(&fiber->std);
+}
+/* }}} */
+
+void zend_fiber_cleanup(void) /* {{{ */
+{
+	zend_fiber *fiber;
+	zend_object *exception = EG(exception);
+	EG(exception) = NULL;
+
+	ZEND_HASH_REVERSE_FOREACH_PTR(&EG(fibers), fiber) {
+		if (fiber->status == ZEND_FIBER_STATUS_SUSPENDED) {
+			fiber->status = ZEND_FIBER_STATUS_SHUTDOWN;
+			GC_ADDREF(&fiber->std);
+			zend_fiber_switch_to(fiber);
+		}
+	} ZEND_HASH_FOREACH_END();
+
+	EG(exception) = exception;
+}
+/* }}} */
+
+/* {{{ proto Fiber::__construct(callable $callback) */
+ZEND_METHOD(Fiber, __construct)
+{
+	zend_fiber *fiber = (zend_fiber *) Z_OBJ_P(getThis());
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_FUNC_EX(fiber->fci, fiber->fci_cache, 0, 0)
+	ZEND_PARSE_PARAMETERS_END();
+
+	// Keep a reference to closures or callable objects until the fiber is started.
+	Z_TRY_ADDREF(fiber->fci.function_name);
+}
+/* }}} */
+
+/* {{{ proto mixed Fiber::start(mixed ...$args) */
+ZEND_METHOD(Fiber, start)
+{
+	zend_fiber *fiber = (zend_fiber *) Z_OBJ_P(getThis());
+
+	if (fiber->status != ZEND_FIBER_STATUS_INIT) {
+		zend_throw_error(zend_ce_fiber_error, "Cannot start a fiber that has already been started");
+		return;
+	}
+
+	fiber->fci.params = ZEND_CALL_ARG(execute_data, 1);
+	fiber->fci.param_count = ZEND_CALL_NUM_ARGS(execute_data);
+
+	if (!zend_fiber_init_context(&fiber->context, zend_fiber_execute, EG(fiber_stack_size))) {
+		zend_throw_error(NULL, "Could not create fiber context");
+		return;
+	}
+
+	fiber->status = ZEND_FIBER_STATUS_RUNNING;
+
+	zend_fiber_switch_to(fiber);
+
+	zval_ptr_dtor(&fiber->fci.function_name);
+
+	if (fiber->status & ZEND_FIBER_STATUS_FINISHED) {
+		RETURN_NULL();
+	}
+
+	RETVAL_COPY_VALUE(&fiber->value);
+	ZVAL_UNDEF(&fiber->value);
+}
+/* }}} */
+
+/* {{{ proto mixed Fiber::suspend(mixed $value) */
+ZEND_METHOD(Fiber, suspend)
+{
+	zend_fiber *fiber = EG(current_fiber);
+	zval *error, *value = NULL;
+
+	if (UNEXPECTED(!fiber)) {
+		zend_throw_error(zend_ce_fiber_error, "Cannot suspend outside of a fiber");
+		return;
+	}
+
+	if (UNEXPECTED(fiber->status != ZEND_FIBER_STATUS_RUNNING)) {
+		if (fiber->status == ZEND_FIBER_STATUS_SHUTDOWN) {
+			zend_throw_error(zend_ce_fiber_error, "Cannot suspend in a force closed fiber");
+		} else {
+			zend_throw_error(zend_ce_fiber_error, "Cannot suspend in a fiber that is not running");
+		}
+		return;
+	}
+
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_ZVAL(value);
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (value) {
+		ZVAL_COPY(&fiber->value, value);
+	} else {
+		ZVAL_NULL(&fiber->value);
+	}
+
+	fiber->execute_data = execute_data;
+	fiber->status = ZEND_FIBER_STATUS_SUSPENDED;
+
+	// Remove running reference while suspended.
+	GC_DELREF(&fiber->std);
+
+	zend_fiber_suspend(fiber);
+
+	if (fiber->status == ZEND_FIBER_STATUS_SHUTDOWN) {
+		// This occurs when the fiber is GC'ed while suspended, do not add a ref.
+		if (EG(fiber_error)) {
+			zend_throw_unwind_exit();
```

Comment:
Add a comment that we use unwind_exit here to not run finally?

Though I'm thinking that it might make sense to merge unwind_exit and finally_exit and just have a flag on them that determines whether or not finally should be executed. Otherwise we'll get problems if/when the behavior of unwind_exit changes.

Reply:
Instead of changing unwind_exit, should we name fiber_exit something more generic, like graceful_exit or finally_exit and use that if/when we want to change the `exit` behavior? Then extension users of unwind_exit won't have behavioral changes.